### ISSUE TITLE
Handle NetworkThread in the crate

### ISSUE THREAD

**Issue by PierreZ** [foundationdb-rs/foundationdb-rs#195]
The networkThread is a major complexity to handle for our users. Official bindings are handling it directly in the library, we could do that as well

**Comment by PierreZ** [comment]
I spent some time researching how to finally fix this (and #132), looking at the C API contract, the official Go/Java/Python/Ruby bindings, and our previous attempts (#234, #356). Here is a concrete plan, feedback very welcome.

### The core constraint

The C API allows `fdb_select_api_version` and `fdb_setup_network` once per process, `fdb_run_network` blocks on a dedicated thread and can never be called twice, and once `fdb_stop_network` has been called the network can never be restarted (error 2025). The docs also require stop + join before a normal process exit.

That "never restart" rule is why stop-on-last-guard-drop cannot fix #132, whether it is today's `NetworkAutoStop` or the refcount in #356: as soon as the count hits zero between two tests, any later `Database::new` fails with 2025. The conclusion I came to: the network belongs to the process, not to a guard.

### What official bindings do

| binding | version selected twice | network start | network stop |
|---|---|---|---|
| Go | same: Ok, different: 2201 | lazy, on first `OpenDatabase` | never |
| Python | same: Ok, different: raise | lazy, on `fdb.open()` | `atexit` handler: stop + join |
| Java | same: Ok, different: throw | lazy, on `fdb.open()` | JVM shutdown hook (removable) |
| Ruby | same: Ok, different: raise | lazy, on `FDB.open` | `at_exit` block: stop + join |

None of them stop the network while the process is alive, and all of them treat re-selecting the same version as a no-op.

### Proposed design: a process-global state machine

The whole lifecycle becomes an explicit state machine in the code. One enum, variants owning their data, transitions as methods doing an exhaustive `match` on the current state, and every FFI call living inside exactly one transition. No more scattered flags (the `VERSION_SELECTED` atomic disappears), and adding a state or an event becomes a compile error until every `match` handles it. The transition table itself gets unit tests.

```rust
enum NetworkLifecycle {
    Uninitialized,
    ApiVersionSelected { api_version: i32 },
    Running { api_version: i32, handle: JoinHandle<()> },
    Stopped { api_version: i32 },
}

static NETWORK: Mutex<NetworkLifecycle> = Mutex::new(NetworkLifecycle::Uninitialized);

impl NetworkLifecycle {
    fn select_api_version(&mut self, version: i32) -> FdbResult<()>;
    fn set_network_option(&mut self, option: NetworkOption) -> FdbResult<()>;
    fn start_network(&mut self) -> FdbResult<()>;
    fn stop_network(&mut self) -> FdbResult<()>;
}
```

All public entry points (`boot`, `FdbApiBuilder`, `NetworkBuilder`, lazy boot in `Database::new`) route through it:

```text
Uninitialized
  |- select_api_version(v) -----------> ApiVersionSelected { v }   (2203 if unsupported)
  '- start_network(): selects default version first, then continues below

ApiVersionSelected { v }
  |- select_api_version(same v): Ok (idempotent) / different: Err(2201)
  |- set_network_option(opt): applied immediately (before setup, as the C API requires)
  '- start_network(): fdb_setup_network under the lock, register atexit hook,
                      spawn "fdb-network" thread ---> Running { v, JoinHandle }

Running { v, handle }
  |- select_api_version(same v): Ok / different: Err(2201)
  |- set_network_option: Err(2009 network_already_setup)
  |- start_network(): Ok (idempotent no-op)
  '- stop_network() or atexit handler: fdb_stop_network + join ---> Stopped { v }

Stopped { v }   (terminal)
  '- start_network() / Database::new: Err(2025 network_cannot_be_restarted)
```

Notes:
- `fdb_setup_network` runs synchronously under the lock before the thread is spawned, so the setup/run ordering is guaranteed without the current condvar dance. Futures simply do not complete until `fdb_run_network` is live, which the C API allows.
- The network thread no longer panics if `fdb_run_network` returns an error: it logs to stderr and stores the error for inspection, like Java/Python/Ruby.
- The atexit handler (registered when the network starts, so it runs before libfdb's own exit-time cleanup) does stop + join, satisfying the documented exit contract. It is panic-free and mirrors the abort-on-stop-failure policy of today's `NetworkAutoStop::drop`.

### Public API changes

```rust
// before
let _guard = unsafe { foundationdb::boot() };            // panics, second call panics too
// after
let _guard = foundationdb::boot()?;                      // safe, idempotent, Result as asked in #132
```

- `boot()` is safe, idempotent, and returns `FdbResult<NetworkAutoStop>`. The guard is kept for compatibility but is now inert: dropping it does nothing, the network keeps running until process exit.
- Lazy boot: `Database::new` / `Database::new_compat` start the network with the default API version if nothing was booted yet, exactly like `fdb.open()` in official bindings. Most programs and tests will not need `boot()` at all.
- Configuration is unchanged: `FdbApiBuilder::default().set_runtime_version(710).build()?.set_option(...)?.boot()?`. `FdbApiBuilder::build()` no longer panics on a second call: same version is Ok, different version is Err(2201).
- Shutdown control is at runtime, mirroring Java's shutdown hook model (they register one and expose `disableShutdownHook()`): the atexit hook is always registered when the network starts, and `api::disable_stop_on_exit()` turns it into a no-op for applications that want to own their shutdown sequence. A runtime switch cannot be flipped by Cargo feature unification, and there is a single code path to test.
- For deterministic shutdown there is an explicit, safe `api::stop_network()`: it goes through the state machine (stop + join, terminal), and the atexit handler then finds the network already stopped and does nothing. This replaces `NetworkStop` with one honest function; after calling it, any FDB use fails with 2025.
- Removed: `NetworkRunner`, `NetworkWait`, `NetworkStop` and `NetworkBuilder::build()` (the unmanaged mode). They were the remaining source of `unsafe` and of the drop-before-exit obligation; only our own tests used them.
- The simulation crate is unaffected: `FdbApiBuilder::build()` still works as select-version-only (no network start, no atexit), and simulation databases come from `new_from_pointer`, which bypasses lazy boot.

### What this fixes

- #132: every test can call `boot()` (or just `Database::new`), in any order, in parallel. `--test-threads=1` is no longer needed. As part of the PR I would split one of the mega-tests (`tests/get.rs`) into individual `#[test]` functions as proof.
- No more `unsafe` at bootstrap, and no more undefined behavior if the guard is leaked or dropped early.
- Clear errors instead of panics for every misuse (2201 wrong version, 2009 option after setup, 2025 use after exit-time stop).

### Differences with #356

Same state machine idea, but: no refcounting and no stop on guard drop (the network runs until exit, which is what makes parallel tests reliable), no unmanaged mode kept, explicit `stop_network()` instead of drop semantics, and the review feedback applied (no stdout prints, no bare unwraps, better naming).

This is a breaking change, so it would land in the next 0.x (release-plz will handle the bump from the `feat!:` commit).

Feedback welcome before I start the implementation.
